File: app/src/main/java/com/silvertak/relationshipsmanager/adapter/SelectablePersonListAdapter.java
package com.silvertak.relationshipsmanager.adapter;

import android.content.Context;
import android.os.Handler;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;

import androidx.annotation.NonNull;
import androidx.databinding.ObservableArrayList;
import androidx.recyclerview.widget.RecyclerView;

import com.silvertak.relationshipsmanager.R;
import com.silvertak.relationshipsmanager.customInterface.OnContactInfoClick;
import com.silvertak.relationshipsmanager.data.PersonRelationshipInfo;
import com.silvertak.relationshipsmanager.databinding.SelectablePersonItemBinding;
import com.silvertak.relationshipsmanager.library.StringLib;

public class SelectablePersonListAdapter extends RecyclerView.Adapter<SelectablePersonListAdapter.CustomViewHolder> implements RecyclerView.OnItemTouchListener{

    private ObservableArrayList<PersonRelationshipInfo> contactInfos = new ObservableArrayList<>();
    private ObservableArrayList<PersonRelationshipInfo> selectInfos = new ObservableArrayList<>();
    private OnContactInfoClick contactInfoClickListener;
    private String strSearchText = "";
    private Context mContext;

    public SelectablePersonListAdapter(Context context){
        mContext = context;
    }

    public void setContactInfos(ObservableArrayList<PersonRelationshipInfo> infos)
    {
        this.contactInfos = infos;
    }

    public void setSelectInfos(ObservableArrayList<PersonRelationshipInfo> infos)
    {
        this.selectInfos = infos;
    }

    public void executeSearch(String strSearchText)
    {
        this.strSearchText = strSearchText;
        notifyDataSetChanged();
    }

    private boolean isSearchTextContain(PersonRelationshipInfo info)
    {
        if(info.getContactInfo() == null
                || StringLib.isEmpty(info.getContactInfo().getName())
                || StringLib.isEmpty(info.getContactInfo().getPhoneNumber()))
            return false;

        if(info.getContactInfo().getName().contains(this.strSearchText)
                || info.getContactInfo().getPhoneNumber().contains(this.strSearchText))
            return true;
        else return false;
    }

    private boolean isSearchMode()
    {
        if(StringLib.isEmpty(this.strSearchText)) return false;
        else return true;
    }

    private int getSearchCount()
    {
        int nResult = 0;
        for(PersonRelationshipInfo info : contactInfos)
            if(isSearchTextContain(info)) nResult++;
        return nResult;
    }

    private PersonRelationshipInfo getNextData(int position) {
        int nSearchCount = 0;
        PersonRelationshipInfo resultInfo = null;
        for (int i = 0; i < contactInfos.size(); i++)
        {
            PersonRelationshipInfo info = contactInfos.get(i);
            if(isSearchTextContain(info))
            {
                if(position == nSearchCount) {resultInfo = info; break;}
                nSearchCount++;
            }
        }
        return resultInfo;
    }

    @Override
    public void onBindViewHolder(@NonNull final CustomViewHolder holder, final int position)
    {
        PersonRelationshipInfo info = (isSearchMode()) ? getNextData(position) : contactInfos.get(position);
        holder.bind(info, mContext);
        holder.mBinding.cardview.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                PersonRelationshipInfo info = contactInfos.get(position);
                if(contactInfos.get(position).getSelected().getValue())
                {
                    info.setSelected(false);
                    holder.setSelectValue(false);
                    if(selectInfos.contains(info)) selectInfos.remove(info);
                }
                else
                {
                    info.setSelected(true);
                    holder.setSelectValue(true);
                    selectInfos.add(info);
                }
            }
        });
        //if(!isSearchMode())
            setAnimation(holder.itemView);
    }

    private void setAnimation(View view) {
        AlphaAnimation anim = new AlphaAnimation(0.0f, 1.0f);
        anim.setDuration(300);
        view.startAnimation(anim);
    }

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        SelectablePersonItemBinding selectablePersonItemBinding = SelectablePersonItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new CustomViewHolder(selectablePersonItemBinding);
    }

    @Override
    public int getItemCount() {
        if(isSearchMode())
            return getSearchCount();
        else
            return contactInfos.size();
    }

    @Override
    public boolean onInterceptTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) {
        return true;
    }

    @Override
    public void onTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) {
    }

    @Override
    public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {
    }

    class CustomViewHolder extends RecyclerView.ViewHolder
    {
        private SelectablePersonItemBinding mBinding;

        public CustomViewHolder(SelectablePersonItemBinding binding)
        {
            super(binding.getRoot());
            this.mBinding = binding;
        }

        void bind(PersonRelationshipInfo info, Context context)
        {
            mBinding.setPersonRelationshipInfo(info);
            initBackgroundColor(info, context);
        }

        void initBackgroundColor(PersonRelationshipInfo info, Context context)
        {
            if(info.getSelected().getValue())
                mBinding.backgroundLayout.setBackgroundColor(context.getResources().getColor(R.color.hashtag));
            else
                mBinding.backgroundLayout.setBackgroundColor(context.getResources().getColor(R.color.fontColor));
        }

        void setSelectValue(boolean bValue)
        {
            PersonRelationshipInfo info = mBinding.getPersonRelationshipInfo();
            info.setSelected(bValue);
            mBinding.setPersonRelationshipInfo(info);
        }
    }
}

File: app/src/main/java/com/silvertak/relationshipsmanager/adapter/SelectablePersonListAdapter.kt
package com.silvertak.relationshipsmanager.adapter

import android.content.Context
import android.os.Build
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import androidx.annotation.RequiresApi
import androidx.databinding.ObservableArrayList
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnItemTouchListener
import com.silvertak.relationshipsmanager.R
import com.silvertak.relationshipsmanager.customInterface.OnContactInfoClick
import com.silvertak.relationshipsmanager.vo.PersonRelationshipInfo
import com.silvertak.relationshipsmanager.databinding.SelectablePersonItemBinding
import com.silvertak.relationshipsmanager.library.StringLib

class SelectablePersonListAdapter(private val mContext: Context) : RecyclerView.Adapter<SelectablePersonListAdapter.CustomViewHolder>(), OnItemTouchListener {
    private var contactInfos = ObservableArrayList<PersonRelationshipInfo>()
    private var selectInfos = ObservableArrayList<PersonRelationshipInfo>()
    private val contactInfoClickListener: OnContactInfoClick? = null
    private var strSearchText = ""
    fun setContactInfos(infos: ObservableArrayList<PersonRelationshipInfo>) {
        contactInfos = infos
    }

    fun setSelectInfos(infos: ObservableArrayList<PersonRelationshipInfo>) {
        selectInfos = infos
    }

    fun executeSearch(strSearchText: String) {
        this.strSearchText = strSearchText
        notifyDataSetChanged()
    }

    private fun isSearchTextContain(info: PersonRelationshipInfo): Boolean {
        if (info.contactInfo == null || StringLib.isEmpty(info.contactInfo.name)
                || StringLib.isEmpty(info.contactInfo.phoneNumber)) return false
        return if (info.contactInfo.name.contains(strSearchText)
                || info.contactInfo.phoneNumber.contains(strSearchText)) true else false
    }

    private val isSearchMode: Boolean
        private get() = if (StringLib.isEmpty(strSearchText)) false else true

    private val searchCount: Int
        private get() {
            var nResult = 0
            for (info in contactInfos) if (isSearchTextContain(info)) nResult++
            return nResult
        }

    private fun getNextData(position: Int): PersonRelationshipInfo? {
        var nSearchCount = 0
        var resultInfo: PersonRelationshipInfo? = null
        for (i in contactInfos.indices) {
            val info = contactInfos[i]
            if (isSearchTextContain(info)) {
                if (position == nSearchCount) {
                    resultInfo = info
                    break
                }
                nSearchCount++
            }
        }
        return resultInfo
    }

    override fun onBindViewHolder(holder: CustomViewHolder, position: Int) {
        val info = if (isSearchMode) getNextData(position) else contactInfos[position]
        holder.bind(info, mContext)
        holder.mBinding.cardview.setOnClickListener {
            val info = contactInfos[position]
            if (contactInfos[position].selected.value!!) {
                info.setSelected(false)
                holder.setSelectValue(false)
                if (selectInfos.contains(info)) selectInfos.remove(info)
            } else {
                info.setSelected(true)
                holder.setSelectValue(true)
                selectInfos.add(info)
            }
        }
        //if(!isSearchMode())
        setAnimation(holder.itemView)
    }

    private fun setAnimation(view: View) {
        val anim = AlphaAnimation(0.0f, 1.0f)
        anim.duration = 300
        view.startAnimation(anim)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomViewHolder {
        val selectablePersonItemBinding = SelectablePersonItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CustomViewHolder(selectablePersonItemBinding)
    }

    override fun getItemCount(): Int {
        return if (isSearchMode) searchCount else contactInfos.size
    }

    override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
        return true
    }

    override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {}

    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}
    inner class CustomViewHolder(val mBinding: SelectablePersonItemBinding) : RecyclerView.ViewHolder(mBinding.root) {
        @RequiresApi(Build.VERSION_CODES.M)
        fun bind(info: PersonRelationshipInfo?, context: Context) {
            mBinding.personRelationshipInfo = info
            initBackgroundColor(info, context)
        }
        @RequiresApi(Build.VERSION_CODES.M)
        private fun initBackgroundColor(info: PersonRelationshipInfo?, context: Context) {
            if (info!!.selected.value!!) mBinding.backgroundLayout.setBackgroundColor(context.getColor(R.color.hashtag))
            else mBinding.backgroundLayout.setBackgroundColor(context.getColor(R.color.fontColor))
        }

        fun setSelectValue(bValue: Boolean) {
            mBinding.personRelationshipInfo.setSelected(bValue)
        }

    }

}